staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/cmd.go

```diff
@@ -266,11 +266,17 @@ func Run(g *Generator) {
 		p := outputPackage.(*protobufPackage)
 
 		path := filepath.Join(g.OutputBase, p.ImportPath())
-		outputPath := filepath.Join(g.OutputBase, p.OutputPath())
 		if p.Vendored {
 			path = filepath.Join(g.VendorOutputBase, p.ImportPath())
-			outputPath = filepath.Join(g.VendorOutputBase, p.OutputPath())
 		}
+		// outputPath is the path where protoc-gen-gogo is expected to output
+		// the file for outputPackage. The output path depends on the go_package
+		// option set in the input proto file. If it is a full Go import path
+		// then it will be generated under this import path within the base
+		// output directory. Note that it is not dependend on the input path
+		// itself. All protobufPackages have their import path as go_package
+		// option set.
+		outputPath := filepath.Join(g.OutputBase, p.OutputPath())
```

Comment:
I'm confused why it is correct for this to ignore `Vendored` and `VendorOutputBase`, while the construction of `path` above still honors it

Reply:
`path` is the input path of the current `generated.proto` file to be processed in that loop. It has been generated using the `ExecutePackage` calls in L236 and L239, either under `g.OutputBase` or `g.VendorOutputBase` depending on whether its a vendored package or not. So to calculate the `path` of that `generated.proto`, we have to differentiate if its a vendored package (`p.Vendored == true`) or not.

That `path` is passed to protoc command called in line 282: `protoc --gogo_out=<g.OutputBase> <path>`. That call will generate a new `generate.pb.go` file. Its path is `outputPath`. Previously `outputPath` was dependent on the location of the input `generated.proto` file. Now since all `generated.proto` files have a correct `go_package` option set, the output location is not dependent on the input location anymore. Instead, the output will be generated under `<g.OutputBase>/<go-package-option>/generated.pb.go`, e.g. `<g.OutputBase>/k8s.io/api/admission/v1beta1/generated.pb.go`. 

Whereas previously it was dependent on the input file location: if the `go_package` option is not a full path (i.e. does not contain a `/`) protoc-gen-gogo will generated an output file under `<g.OutputBase>/<input-file-directory>/generated.pb.go`. E.g.: for `protoc --gogo_out=<g.OutputBase> path/to/vendor/k8s.io/api/admission/v1beta1/generated.proto` a file `<g.OutputBase>/path/to/vendor/k8s.io/api/admission/v1beta1/generated.proto`, will be generated if no path as `go_package` is set.

Have you read my comment above? https://github.com/kubernetes/kubernetes/pull/105027#issuecomment-922445525 That shows an example of this behaviour.

Reply:
But `<g.OutputBase>/k8s.io/api/admission/v1beta1/generated.pb.go` is not correct, right? we want `<g.VendorOutputBase>/k8s.io/api/admission/v1beta1/generated.pb.go`